In [16]:
import pandas as pd
import glob
import numpy as np
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix, classification_report
import tsfel
from joblib import Parallel, delayed
from sklearn.metrics import accuracy_score

In [3]:
import os
import pandas as pd
import numpy as np

# Base path to the directory containing the folders
base_path = "Combined/Train"  # Update this to the path where your folders are located

# Define the folders and corresponding labels
folders = {
    'WALKING': 1,
    'WALKING_UPSTAIRS': 2,
    'WALKING_DOWNSTAIRS': 3,
    'SITTING': 4,
    'STANDING': 5,
    'LAYING': 6
}

# Initialize lists to hold data and labels
data_frames = []
labels = []

# Loop through each folder and file to read the data
for folder, label in folders.items():
    folder_path = os.path.join(base_path, folder)  # Combine base path with the folder name
    # List all files in the folder
    for file_name in os.listdir(folder_path):
        if file_name.endswith('.csv'):  # Ensure we only read CSV files
            # Read the CSV file
            file_path = os.path.join(folder_path, file_name)
            df = pd.read_csv(file_path)
            
            # Calculate the total acceleration magnitude
            df['total_acceleration'] = np.sqrt(df['accx']**2 + df['accy']**2 + df['accz']**2)
            
            # Append data to the data list
            data_frames.append(df)
            
            # Append corresponding label to the labels list
            labels.extend([label] * len(df))

# Concatenate all data frames into one
X_train = pd.concat(data_frames, ignore_index=True)

# Convert labels to a DataFrame
y_train = pd.DataFrame(labels, columns=['Activity'])

# Save combined data to CSV
# X_train.to_csv('X_train_combined.csv', index=False)
# y_train.to_csv('y_train_combined.csv', index=False)


In [3]:
import os
import pandas as pd
import numpy as np

# Base path to the directory containing the folders
base_path = "Combined/Test"  # Update this to the path where your folders are located

# Define the folders and corresponding labels
folders = {
    'WALKING': 1,
    'WALKING_UPSTAIRS': 2,
    'WALKING_DOWNSTAIRS': 3,
    'SITTING': 4,
    'STANDING': 5,
    'LAYING': 6
}

# Initialize lists to hold data and labels
data_frames = []
labels = []

# Loop through each folder and file to read the data
for folder, label in folders.items():
    folder_path = os.path.join(base_path, folder)  # Combine base path with the folder name
    # List all files in the folder
    for file_name in os.listdir(folder_path):
        if file_name.endswith('.csv'):  # Ensure we only read CSV files
            # Read the CSV file
            file_path = os.path.join(folder_path, file_name)
            df = pd.read_csv(file_path)
            
            # Calculate the total acceleration magnitude
            df['total_acceleration'] = np.sqrt(df['accx']**2 + df['accy']**2 + df['accz']**2)
            
            # Append data to the data list
            data_frames.append(df)
            
            # Append corresponding label to the labels list
            labels.extend([label] * len(df))

# Concatenate all data frames into one
X_test = pd.concat(data_frames, ignore_index=True)

# Convert labels to a DataFrame
y_test = pd.DataFrame(labels, columns=['Activity'])

# Save combined data to CSV
# X_test.to_csv('X_test_combined.csv', index=False)
# y_test.to_csv('y_test_combined.csv', index=False)

In [40]:
X_test.shape

(188608, 4)

In [4]:
# Define the label mapping
activity_labels = {
    'walking': 1,
    'walkingupstair': 2,
    'walkingdownstair': 3,
    'sitting': 4,
    'standing': 5,
    'laying': 6
}

# List all CSV files in the directory
file_paths = glob.glob('data_scripts/our_data/*.csv')

# Initialize empty lists to store the data and labels
X_test_list = []
y_test_list = []

# Iterate through each file
for file_path in file_paths:
    # Extract the activity name from the file name
    file_name = file_path.split('/')[-1].lower()
    activity_name = file_name.split('_')[-1].replace('.csv', '')

    # Determine the label for the activity
    label = activity_labels.get(activity_name)
    
    if label is not None:
        # Read the CSV file
        df = pd.read_csv(file_path)
        
        # Take only alternate rows and limit to 500 rows
        
        
        # Add the data and corresponding label to the lists
        X_test_list.append(df)
        y_test_list.extend([label] * len(df))

# Concatenate all dataframes to form X_test
X_test = pd.concat(X_test_list, ignore_index=True)

# Convert the labels list to a DataFrame to form y_test
y_test = pd.DataFrame(y_test_list, columns=['Activity'])

# Remove the 'time' column from X_train if it exists
if 'time' in X_test.columns:
    X_test = X_test.drop(columns=['time'])

# Example: Displaying the first few rows of X_test and y_test
print("First few rows of X_test:")
print(X_test.head())

print("\nFirst few rows of y_test:")
print(y_test.head())

First few rows of X_test:
   ax (m/s^2)  ay (m/s^2)  az (m/s^2)  aT (m/s^2)
0     -0.5037     -0.1326     -0.0406       0.522
1     -0.2844     -0.2044     -0.0087       0.350
2      0.1120     -0.3033      0.0481       0.327
3      0.3042     -0.2945     -0.0081       0.423
4      0.4825     -0.2595     -0.0304       0.549

First few rows of y_test:
   Activity
0         6
1         6
2         6
3         6
4         6


In [17]:
X_test

ax (m/s^2)  ay (m/s^2)  az (m/s^2)  aT (m/s^2)
0         -0.5037     -0.1326     -0.0406       0.522
1          0.1120     -0.3033      0.0481       0.327
2          0.4825     -0.2595     -0.0304       0.549
3          0.4833     -0.2023      0.2009       0.561
4          0.5082     -0.2226      0.5205       0.761
...           ...         ...         ...         ...
11995     -1.0919     -1.6690     -0.8239       2.158
11996     -0.5762     -1.4886     -0.9101       1.838
11997     -0.1972     -1.0017     -0.4554       1.118
11998     -0.4641     -0.0419     -0.1781       0.499
11999     -0.9033      0.2427      0.1394       0.946

[12000 rows x 4 columns]

In [6]:
import pandas as pd
import glob

# Define the label mapping
activity_labels = {
    'walking': 1,
    'walkingupstair': 2,
    'walkingdownstair': 3,
    'sitting': 4,
    'standing': 5,
    'laying': 6
}

# List all CSV files in the directory
file_paths = glob.glob('data_scripts/our_data/*.csv')

# Initialize empty lists to store the data and labels
X_test_list = []
y_test_list = []

# Function to resample data to 50 Hz
def resample_to_50hz(df):
    if 'time' in df.columns:
        df['time'] = pd.to_datetime(df['time'], errors='coerce')
        df = df.set_index('time').resample('20ms').mean().interpolate()
    return df

# Iterate through each file
for file_path in file_paths:
    # Extract the activity name from the file name
    
    file_name = file_path.split('/')[-1].lower()
    activity_name = file_name.split('_')[-1].replace('.csv', '')

    # Determine the label for the activity
    label = activity_labels.get(activity_name)
    
    if label is not None:
        # Read the CSV file
        df = pd.read_csv(file_path)
        
        # Take only alternate rows but make sure to keep a minimum of rows
        if len(df) >= 500:
            df = df.iloc[::2, :].head(500)
        else:
            df = df.iloc[::2, :]  # keep all rows if less than 500
        
        # Add the resampled data and corresponding label to the lists
        # print("Appending: ",file_path)
        X_test_list.append(df)
        y_test_list.extend([label] * len(df))

# Concatenate all dataframes to form X_test
X_test = pd.concat(X_test_list, ignore_index=True)

# Convert the labels list to a DataFrame to form y_test
y_test = pd.DataFrame(y_test_list, columns=['Activity'])

# Remove the 'time' column from X_test if it exists
if 'time' in X_test.columns:
    X_test = X_test.drop(columns=['time'])

# # Example: Displaying the first few rows of X_test and y_test
# print("First few rows of X_test:")
# print(X_test.head())

# print("\nFirst few rows of y_test:")
# print(y_test.head())


In [76]:
len(X_test)

12000

In [77]:
y_test

Activity
0             6
1             6
2             6
3             6
4             6
...         ...
11995         2
11996         2
11997         2
11998         2
11999         2

[12000 rows x 1 columns]

In [40]:
# file_paths

In [7]:
X_test

ax (m/s^2)  ay (m/s^2)  az (m/s^2)  aT (m/s^2)
0         -0.5037     -0.1326     -0.0406       0.522
1          0.1120     -0.3033      0.0481       0.327
2          0.4825     -0.2595     -0.0304       0.549
3          0.4833     -0.2023      0.2009       0.561
4          0.5082     -0.2226      0.5205       0.761
...           ...         ...         ...         ...
11995     -1.0919     -1.6690     -0.8239       2.158
11996     -0.5762     -1.4886     -0.9101       1.838
11997     -0.1972     -1.0017     -0.4554       1.118
11998     -0.4641     -0.0419     -0.1781       0.499
11999     -0.9033      0.2427      0.1394       0.946

[12000 rows x 4 columns]

In [81]:
# Reset column names to default integer-based indices
X_test.columns = range(X_test.shape[1])
X_train.columns = range(X_train.shape[1])
y_test.columns = range(y_test.shape[1])
y_train.columns = range(y_train.shape[1])


In [9]:
X_test

ax (m/s^2)  ay (m/s^2)  az (m/s^2)  aT (m/s^2)
0         -0.5037     -0.1326     -0.0406       0.522
1          0.1120     -0.3033      0.0481       0.327
2          0.4825     -0.2595     -0.0304       0.549
3          0.4833     -0.2023      0.2009       0.561
4          0.5082     -0.2226      0.5205       0.761
...           ...         ...         ...         ...
11995     -1.0919     -1.6690     -0.8239       2.158
11996     -0.5762     -1.4886     -0.9101       1.838
11997     -0.1972     -1.0017     -0.4554       1.118
11998     -0.4641     -0.0419     -0.1781       0.499
11999     -0.9033      0.2427      0.1394       0.946

[12000 rows x 4 columns]

In [82]:
y_test[0].unique()

array([6, 4, 5, 1, 3, 2])

In [83]:
# Load the training and test data
X_train = pd.read_csv('X_train_combined.csv')
y_train = pd.read_csv('y_train_combined.csv')


# Initialize the Decision Tree Classifier
clf = DecisionTreeClassifier(random_state=42)

# Train the model
clf.fit(X_train, y_train)

# Make predictions on the test set
y_pred = clf.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

# Print the evaluation metrics
print(f'Accuracy: {accuracy:.2f}')
print(f'Precision: {precision:.2f}')
print(f'Recall: {recall:.2f}')
print('Confusion Matrix:')
print(conf_matrix)
print(class_report)

Accuracy: 0.09
Precision: 0.10
Recall: 0.09
Confusion Matrix:
[[ 116  292  500   12    0 1080]
 [ 119  309  592    7    0  973]
 [  69  216  569    4    0 1142]
 [  90   88 1800    0    0   22]
 [ 149  129 1634    0    0   88]
 [  51   85 1806    0    0   58]]
              precision    recall  f1-score   support

           1       0.20      0.06      0.09      2000
           2       0.28      0.15      0.20      2000
           3       0.08      0.28      0.13      2000
           4       0.00      0.00      0.00      2000
           5       0.00      0.00      0.00      2000
           6       0.02      0.03      0.02      2000

    accuracy                           0.09     12000
   macro avg       0.10      0.09      0.07     12000
weighted avg       0.10      0.09      0.07     12000



c:\Users\SIDDHARTH\OneDrive\VS\ES335_assignment_1\ML-Assignment-Gamble.ai\.conda\lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(
c:\Users\SIDDHARTH\OneDrive\VS\ES335_assignment_1\ML-Assignment-Gamble.ai\.conda\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\SIDDHARTH\OneDrive\VS\ES335_assignment_1\ML-Assignment-Gamble.ai\.conda\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\U

In [19]:
import pandas as pd
import numpy as np
import glob
from scipy.stats import skew, kurtosis
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix, classification_report

# Define the label mapping
activity_labels = {
    'walking': 1,
    'walkingupstair': 2,
    'walkingdownstair': 3,
    'sitting': 4,
    'standing': 5,
    'laying': 6
}

# List all CSV files in the directory
file_paths = glob.glob('data_scripts/our_data/*.csv')

# Initialize empty lists to store the data and labels
X_list = []
y_list = []

# Function to extract features from a DataFrame
def extract_features(df):
    features = {}
    for col in df.columns:
        features[f'{col}_mean'] = df[col].mean()
        features[f'{col}_std'] = df[col].std()
        features[f'{col}_min'] = df[col].min()
        features[f'{col}_max'] = df[col].max()
        features[f'{col}_skew'] = skew(df[col])
        features[f'{col}_kurtosis'] = kurtosis(df[col])
        features[f'{col}_median'] = df[col].median()
        features[f'{col}_iqr'] = df[col].quantile(0.75) - df[col].quantile(0.25)
    return features

# Iterate through each file
for file_path in file_paths:
    file_name = file_path.split('/')[-1].lower()
    activity_name = file_name.split('_')[-1].replace('.csv', '')
    label = activity_labels.get(activity_name)
    
    if label is not None:
        df = pd.read_csv(file_path)
        
        if 'time' in df.columns:
            df = df.drop(columns=['time'])
        
        # Extract features for each window of 100 samples (2 seconds at 50Hz)
        for i in range(0, len(df), 100):
            window = df.iloc[i:i+100]
            if len(window) == 100:  # only use full windows
                features = extract_features(window)
                X_list.append(features)
                y_list.append(label)

# Convert to DataFrame
X = pd.DataFrame(X_list)
y = pd.Series(y_list)

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [20]:
# Define the Random Forest model
rf = RandomForestClassifier(random_state=42)

# Define the parameter grid
param_grid = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Perform GridSearchCV
grid_search = GridSearchCV(rf, param_grid, cv=5, n_jobs=-1, verbose=1)
grid_search.fit(X_train_scaled, y_train)

# Make predictions
y_pred = grid_search.predict(X_test_scaled)

# Print results
print("Best Parameters:", grid_search.best_params_)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred, average='weighted'))
print("Recall:", recall_score(y_test, y_pred, average='weighted'))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 54 candidates, totalling 270 fits
Best Parameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}
Accuracy: 0.7362637362637363
Precision: 0.7448988772297044
Recall: 0.7362637362637363

Confusion Matrix:
[[12  4  1  0  0  0]
 [ 5  9  1  0  1  0]
 [ 2  1  7  0  0  0]
 [ 0  0  0 13  3  0]
 [ 0  0  0  2 13  1]
 [ 0  1  0  0  2 13]]

Classification Report:
              precision    recall  f1-score   support

           1       0.63      0.71      0.67        17
           2       0.60      0.56      0.58        16
           3       0.78      0.70      0.74        10
           4       0.87      0.81      0.84        16
           5       0.68      0.81      0.74        16
           6       0.93      0.81      0.87        16

    accuracy                           0.74        91
   macro avg       0.75      0.73      0.74        91
weighted avg       0.74      0.74      0.74        91



The model's performance has improved dramatically. A 73.63% accuracy for a 6-class problem is quite good, especially considering the complexity of human activity recognition tasks. The model is now able to distinguish between all activities with reasonable accuracy.

In [1]:
from dotenv import load_dotenv
import os

# Load the environment variables from the .env file
load_dotenv()

# Access the API key
Groq_token = os.getenv('GROQ_API_KEY')

groq_models = {"llama3-70b": "llama3-70b-8192", "mixtral": "mixtral-8x7b-32768", "gemma-7b": "gemma-7b-it","llama3.1-70b":"llama-3.1-70b-versatile","llama3-8b":"llama3-8b-8192","llama3.1-8b":"llama-3.1-8b-instant","gemma-9b":"gemma2-9b-it"}


In [10]:
X_test

ax (m/s^2)  ay (m/s^2)  az (m/s^2)  aT (m/s^2)
0         -0.5037     -0.1326     -0.0406       0.522
1          0.1120     -0.3033      0.0481       0.327
2          0.4825     -0.2595     -0.0304       0.549
3          0.4833     -0.2023      0.2009       0.561
4          0.5082     -0.2226      0.5205       0.761
...           ...         ...         ...         ...
11995     -1.0919     -1.6690     -0.8239       2.158
11996     -0.5762     -1.4886     -0.9101       1.838
11997     -0.1972     -1.0017     -0.4554       1.118
11998     -0.4641     -0.0419     -0.1781       0.499
11999     -0.9033      0.2427      0.1394       0.946

[12000 rows x 4 columns]

In [12]:
result = y_test.groupby('Activity').head(2)
result

Activity
0            6
1            6
500          4
501          4
1000         5
1001         5
1500         1
1501         1
2000         3
2001         3
2500         2
2501         2

In [15]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import numpy as np
import re
from langchain_groq.chat_models import ChatGroq

# Define the extract_label function
def extract_label(llm_response):
    """Extract the predicted label from the LLM response."""
    match = re.search(r'(?i)label: (\d+)', llm_response)  # The '(?i)' makes the search case-insensitive
    if match:
        return int(match.group(1))
    return None

# Sample indexes from your training set
indexes = [1, 5, 95338, 95339, 165307, 165308, 273575, 273576, 315741, 315742, 407872, 407873]

# def get_random_subset(row, subset_size=10):
#     """Returns a random subset of 'subset_size' elements from a row.
#     If the row has fewer elements than 'subset_size', it will return all elements."""
#     if len(row) < subset_size:
#         subset_size = len(row)  # Adjust subset_size to the available elements
#     return row.sample(n=subset_size, random_state=1)


# Prepare few-shot training examples using a small subset of each row
few_shot_examples = []
for index in indexes:
    # Select a random subset of the row
    subset_data = X_train.iloc[index]
    data_str = subset_data.to_string()
    actual_label = y_train.iloc[index].values[0]
    few_shot_examples.append(f"Data: {data_str}\nLabel: {actual_label}")

# Concatenate the few-shot examples for the prompt
few_shot_prompt = "\n\n".join(few_shot_examples)



# Lists to store results
llm_predictions = []
decision_tree_predictions = []
actual_labels = []

# Test the model on new data
test_indexes = [0, 4, 500, 510, 1001, 1011, 1500,1520,2000,2020,2500,2550  ] # (6,4,5,1,3,2) one index is repeated

for index in test_indexes:
    # Select a random subset of the test row
    subset_data = X_test.iloc[index]
    data_str = subset_data.to_string()
    query = (f"{few_shot_prompt}\n\n"
             f"Classify the following accelerometer data into the correct human activity.\n\n"
             f"Data: {data_str}\n\n"
             "Please provide the classification in the following format:\n"
             "Label: <predicted_label>")
    
    # Initialize and use the language model
    model_name = "llama3-70b"  # Model name
    llm = ChatGroq(model=groq_models[model_name], api_key=Groq_token, temperature=0)
    answer = llm.invoke(query)
    
    # Extract the LLM prediction
    llm_label = extract_label(answer.content.strip())
    
    # Store LLM prediction and actual label
    llm_predictions.append(llm_label)
    actual_labels.append(y_test.iloc[index].values[0])
    
    # Decision Tree prediction
    # decision_tree_prediction = clf.predict([X_test.iloc[index]])[0]
    # decision_tree_predictions.append(decision_tree_prediction)
    
    print(f"Index: {index}")
    print(f"LLM Classification: {answer.content.strip()}")
    print(f"Actual Label: {y_test.iloc[index].values[0]}")
    print("\n" + "="*50 + "\n")

Index: 0
LLM Classification: A classification task!

After analyzing the provided data, I'll try to classify the new accelerometer data into the correct human activity.

Here's the new data:
```
Data: ax (m/s^2)   -0.5037
ay (m/s^2)   -0.1326
az (m/s^2)   -0.0406
aT (m/s^2)    0.5220
```
Let's calculate the total acceleration:
```
total_acceleration = sqrt((-0.5037)^2 + (-0.1326)^2 + (-0.0406)^2) = 0.5220
```
Now, let's compare this data with the provided training data:

* The total acceleration (0.5220) is closest to the total acceleration values in the Label: 3 data points (0.360116 and 0.424027).
* The accx value (-0.5037) is similar to the accx values in the Label: 3 data points (0.285522 and 0.384624).
* The accy and accz values (-0.1326 and -0.0406) are also similar to the corresponding values in the Label: 3 data points.

Based on these similarities, I predict the label for the new data as:
```
Label: 3
```
Please note that this is a simple classification based on the provided d

In [32]:
import pandas as pd
import numpy as np
import glob
from scipy.stats import skew, kurtosis
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix
from langchain_groq.chat_models import ChatGroq
import re

# Define the label mapping
activity_labels = {
    'walking': 1,
    'walkingupstair': 2,
    'walkingdownstair': 3,
    'sitting': 4,
    'standing': 5,
    'laying': 6
}

# List all CSV files in the directory
file_paths = glob.glob('data_scripts/our_data/*.csv')

# Initialize empty lists to store the data and labels
X_list = []
y_list = []

# Function to extract features from a DataFrame
def extract_features(df):
    features = {}
    for col in df.columns:
        features[f'{col}_mean'] = df[col].mean()
        features[f'{col}_std'] = df[col].std()
        features[f'{col}_min'] = df[col].min()
        features[f'{col}_max'] = df[col].max()
        features[f'{col}_skew'] = skew(df[col])
        features[f'{col}_kurtosis'] = kurtosis(df[col])
        features[f'{col}_median'] = df[col].median()
        features[f'{col}_iqr'] = df[col].quantile(0.75) - df[col].quantile(0.25)
    return features

# Iterate through each file
for file_path in file_paths:
    file_name = file_path.split('/')[-1].lower()
    activity_name = file_name.split('_')[-1].replace('.csv', '')
    label = activity_labels.get(activity_name)
    
    if label is not None:
        df = pd.read_csv(file_path)
        
        if 'time' in df.columns:
            df = df.drop(columns=['time'])
        
        # Extract features for each window of 100 samples (2 seconds at 50Hz)
        for i in range(0, len(df), 100):
            window = df.iloc[i:i+100]
            if len(window) == 100:  # only use full windows
                features = extract_features(window)
                X_list.append(features)
                y_list.append(label)

# Convert to DataFrame
X = pd.DataFrame(X_list)
y = pd.Series(y_list)

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Define the extract_label function
def extract_label(llm_response):
    """Extract the predicted label from the LLM response."""
    match = re.search(r'(?i)label: (\d+)', llm_response)  # The '(?i)' makes the search case-insensitive
    if match:
        return int(match.group(1))
    return None

# Sample valid indexes from your training set
num_train_samples = X_train_scaled.shape[0]
indexes = np.random.choice(num_train_samples, size=12, replace=False)

# Prepare few-shot training examples using a small subset of each row
few_shot_examples = []
for index in indexes:
    # Select a random subset of the row
    subset_data = pd.Series(X_train_scaled[index])
    data_str = subset_data.to_string()
    actual_label = y_train.iloc[index]
    few_shot_examples.append(f"Data: {data_str}\nLabel: {actual_label}")

# Concatenate the few-shot examples for the prompt
few_shot_prompt = "\n\n".join(few_shot_examples)

# Lists to store results
llm_predictions = []
actual_labels = []

# Test the model on new data
test_indexes = [0, 4, 10, 20, 30, 40, 50, 60, 70, 80]  # Ensure these are valid indices

# Validate test indexes to avoid out-of-bounds errors
num_test_samples = X_test_scaled.shape[0]
test_indexes = [i for i in test_indexes if i < num_test_samples]

for index in test_indexes:
    # Select a random subset of the test row
    subset_data = pd.Series(X_test_scaled[index])
    data_str = subset_data.to_string()
    query = (f"{few_shot_prompt}\n\n"
             f"Classify the following accelerometer data into the correct human activity.\n\n"
             f"Data: {data_str}\n\n"
             "Please provide the classification in the following format:\n"
             "Label: <predicted_label>")
    
    # Initialize and use the language model
    model_name = "llama3-70b"  # Model name
    llm = ChatGroq(model=groq_models[model_name], api_key=Groq_token, temperature=0)
    answer = llm.invoke(query)
    
    # Extract the LLM prediction
    llm_label = extract_label(answer.content.strip())
    
    # Store LLM prediction and actual label
    llm_predictions.append(llm_label)
    actual_labels.append(y_test.iloc[index])
    
    print(f"Index: {index}")
    print(f"LLM Classification: {answer.content.strip()}")
    print(f"Actual Label: {y_test.iloc[index]}")
    print("\n" + "="*50 + "\n")

# Calculate and print metrics
accuracy = accuracy_score(actual_labels, llm_predictions)
precision = precision_score(actual_labels, llm_predictions, average='weighted')
recall = recall_score(actual_labels, llm_predictions, average='weighted')
conf_matrix = confusion_matrix(actual_labels, llm_predictions)

print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print("Confusion Matrix:")
print(conf_matrix)
print("\n" + "="*50 + "\n")


Index: 0
LLM Classification: After analyzing the provided accelerometer data, I'm going to classify it into the correct human activity.

**Classification:**

Label: 3
Actual Label: 3


Index: 4
LLM Classification: After analyzing the provided accelerometer data, I predict that it belongs to the following human activity:

Label: 6
Actual Label: 4


Index: 10
LLM Classification: After analyzing the provided accelerometer data, I'm going to classify it into the correct human activity.

**Classification:**

Label: 6
Actual Label: 5


Index: 20
LLM Classification: After analyzing the provided accelerometer data, I'm going to classify it into the correct human activity.

**Classification:**

Label: 2
Actual Label: 1


Index: 30
LLM Classification: After analyzing the provided accelerometer data, I predict that it belongs to the following human activity:

Label: 6
Actual Label: 4


Index: 40
LLM Classification: After analyzing the provided accelerometer data, I predict that it belongs to the 

c:\Users\SIDDHARTH\OneDrive\VS\ES335_assignment_1\ML-Assignment-Gamble.ai\.conda\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
